src/main/java/uk/gov/justice/digital/delius/service/UserAccessService.java

```diff
@@ -0,0 +1,73 @@
+package uk.gov.justice.digital.delius.service;
+
+import lombok.AllArgsConstructor;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.security.access.AccessDeniedException;
+import org.springframework.security.core.GrantedAuthority;
+import org.springframework.stereotype.Service;
+import uk.gov.justice.digital.delius.data.api.AccessLimitation;
+import uk.gov.justice.digital.delius.helpers.CurrentUserSupplier;
+
+import java.util.Collection;
+import java.util.Set;
+
+@Service
+@AllArgsConstructor
+public class UserAccessService {
+    private final UserService userService;
+    private final OffenderService offenderService;
+    private final CurrentUserSupplier currentUserSupplier;
+    @Value("${user-access.scopes.dont-apply-exclusions-for}")
+    private final Set<String> ignoreExclusionRoles;
+    @Value("${user-access.scopes.dont-apply-restrictions-for}")
+    private final Set<String> ignoreRestrictionRoles;
+
+    public void checkExclusionsAndRestrictions(String crn, Collection<? extends GrantedAuthority> authorities) {
+        final var username = currentUserSupplier.username();
+        final var offender = offenderService.getOffenderByCrn(crn);
+        if (username.isPresent() && shouldCheckExclusion(authorities)) {
+
+            final var excludedException = offender
+                .map(o -> userService.accessLimitationOf(username.get(), o))
+                .filter(AccessLimitation::isUserExcluded)
+                .map(accessLimitation -> new AccessDeniedException(accessLimitation.getExclusionMessage()));
+
+            if (excludedException.isPresent())
+                throw excludedException.get();
+        }
+
+        if (shouldCheckRestriction(authorities)) {
+            final var restrictedException = offender
+                .map(o -> username.map(u -> userService.accessLimitationOf(u, o))
```

Comment:
can't username be null here?

Reply:
ah, so username is an Optional ...

Reply:
yeah, slightly funny, not hugely enamoured with the nested optionals but it was a sufficiently simple flow that I was happy with it. Could extract a `getAccessLimitationOf` method or similar if you think that would read better?

Reply:
Actually - better still I noticed there's precedent for using vavr.Either - think that might read better.

Reply:
Yeah, we have started using Either on new bits of code, I much prefer it to "flow by exception"

Reply:
Ignore my previous comment - I'm getting confused about how Either works!  Re using it in place of flow by exception I did consider doing something similar but then I ended up having to build my own responses in the controller and I figured that would add extra overhead for anyone wanting to reuse this code in future - and having the controller advice handle it seemed like a sufficiently idiomatic and well understood mechanism that I decided to keep it simple.